### PR TITLE
Remove Arrow dependency from the `datasource.hpp` public header

### PR DESCRIPTION
**Pull request by vuule**
## Description
Remove arrow dependency from `datasource.hpp`.

Breaking only because users of `arrow_io_source` now need to include the new `arrow_io_source.hpp` header instead on `datasource.hpp`

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vuule** [comment]
Will re-target for 23.10 once changes are done; this one is not worth the risk to merge late into the release cycle.

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
Thanks @vuule ! LGTM

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/include/cudf/io/arrow_io_source.hpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+// We disable warning 2810 to workaround the compile issue (warning treated as error):
+// result.h(263): error #2810-D: ignoring return value type with "nodiscard" attribute
+#ifdef __CUDACC__
+#pragma nv_diag_suppress 2810
+#endif
+#include <arrow/result.h>
+#ifdef __CUDACC__
+#pragma nv_diag_default 2810
+#endif
```
Comment (ttnghia) [inline cpp/include/cudf/io/arrow_io_source.hpp:43]: We've recently upgraded to arrow 12. Can you try removing these suppression please?

Comment (ttnghia) [reply cpp/include/cudf/io/arrow_io_source.hpp]: Testing locally your latest changes in https://github.com/rapidsai/cudf/pull/13698/commits/031ca64f5fab857befc650ed61e0d12aa2d1bf73.

File: cpp/include/cudf/io/arrow_io_source.hpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "datasource.hpp"
+
+#include <arrow/buffer.h>
+
```
Comment (ttnghia) [inline cpp/include/cudf/io/arrow_io_source.hpp:22]:
```suggestion
```

Comment (ttnghia) [reply cpp/include/cudf/io/arrow_io_source.hpp:22]: In addition, my clang-format changes this file a lot. So I wonder why CI clang-format doesn't complain?

File: cpp/include/cudf/io/arrow_io_source.hpp (hunk 1 above)
Comment (ttnghia) [reply cpp/include/cudf/io/arrow_io_source.hpp]: The test build was successful. So we may be safe to remove such suppressions.

Comment (vuule) [reply cpp/include/cudf/io/arrow_io_source.hpp]: Let's see what CI says :)

File: python/cudf/cudf/_lib/cpp/io/arrow_io_source.pxd [hunk 3]
```diff
@@ -11,5 +11,5 @@ cdef extern from "cudf/io/arrow_io_source.hpp" \
         namespace "cudf::io" nogil:
 
     cdef cppclass arrow_io_source(cudf_io_datasource.datasource):
-        arrow_io_source(string arrow_uri) except +
+        arrow_io_source(const string& arrow_uri) except +
```
Comment (ttnghia) [inline python/cudf/cudf/_lib/cpp/io/arrow_io_source.pxd:14]: east const.

Comment (vuule) [reply python/cudf/cudf/_lib/cpp/io/arrow_io_source.pxd:14]: For some reason I only see "const string" in pxd files so I kept this consistent with other instances. I'm not sure if our east const guideline reached Cython.

**Review by ttnghia — APPROVED** [review]

File: cpp/include/cudf/io/arrow_io_source.hpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    [[nodiscard]] uint8_t const* data() const override { return arrow_buffer->data(); }
+  };
+
+ public:
+  /**
+   * @brief Constructs an object from an Apache Arrow Filesystem URI
+   *
+   * @param arrow_uri Apache Arrow Filesystem URI
+   */
+  explicit arrow_io_source(std::string const& arrow_uri)
```
Comment (PointKernel) [inline cpp/include/cudf/io/arrow_io_source.hpp:62]: Same to all other member functions. Maybe considering separating class declaration from definition

**Review by PointKernel — APPROVED** [review]
Just one small software engineering comment. LGTM

File: cpp/include/cudf/io/arrow_io_source.hpp (hunk 4 above)
Comment (vuule) [reply cpp/include/cudf/io/arrow_io_source.hpp:62]: I want to keep this class separate from libcudf, but we don't have to take that route.
@davidwendt thoughts on keeping the implementation in libcudf?

Comment (davidwendt) [reply cpp/include/cudf/io/arrow_io_source.hpp:62]: I'm not sure I understand what you mean by keeping this separate from libcudf.
I think Yunsong means moves the implementation to `cpp/src/io/arrow_io_source.cpp` and keep only the declarations in this header file. I agree with that approach.

File: python/cudf/cudf/_lib/cpp/io/arrow_io_source.pxd (hunk 3 above)
Comment (vyasr) [reply python/cudf/cudf/_lib/cpp/io/arrow_io_source.pxd:14]: Cython doesn't support east const.

File: cpp/include/cudf/io/arrow_io_source.hpp (hunk 4 above)
Comment (vuule) [reply cpp/include/cudf/io/arrow_io_source.hpp:62]: I meant that I will need to compile cpp/src/io/arrow_io_source.cpp as part of libcudf. Not an issue, just different from what I imagined in the beginning.

File: cpp/src/io/utilities/arrow_io_source.cpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace cudf::io {
+
+/**
+ * @brief Implementation for an owning buffer where `arrow::Buffer` holds the data.
+ */
+class arrow_io_buffer : public datasource::buffer {
+  std::shared_ptr<arrow::Buffer> arrow_buffer;
+
+ public:
```
Comment (robertmaynard) [inline cpp/src/io/utilities/arrow_io_source.cpp:34]: since you have public inheritance this `public:` is a no-op

**Review by robertmaynard — APPROVED** [review]

**Pull request merged**